src/FakeItEasy/Core/CallOrderingContext.cs

```diff
@@ -0,0 +1,87 @@
+﻿namespace FakeItEasy.Core
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    internal class CallOrderingContext : ICallOrderingContext
+    {
+        private readonly CallWriter callWriter;
+        private readonly HashSet<FakeManager> fakeManagers;
+        private readonly List<AssertedCall> assertedCalls;
+        private int currentSequenceNumber;
+
+        public CallOrderingContext(CallWriter callWriter)
+        {
+            this.callWriter = callWriter;
+            this.currentSequenceNumber = -1;
+            this.fakeManagers = new HashSet<FakeManager>();
+            this.assertedCalls = new List<AssertedCall>();
+        }
+
+        public void CheckNextCall(FakeManager fakeManager, Func<IFakeObjectCall, bool> callPredicate, string callDescription, Repeated repeatConstraint)
+        {
+            this.fakeManagers.Add(fakeManager);
+            this.assertedCalls.Add(new AssertedCall { CallDescription = callDescription, RepeatDescription = repeatConstraint.ToString() });
+            var allCalls = this.fakeManagers.SelectMany(f => f.AllRecordedCalls.Cast<IFakeObjectCall>()).OrderBy(c => c.SequenceNumber).ToList();
+            var calls = allCalls
+                .Where(callPredicate)
+                .SkipWhile(c => c.SequenceNumber <= this.currentSequenceNumber)
+                .ToList();
+
+            if (!repeatConstraint.Matches(calls.Count))
+            {
+                ThrowExceptionWhenAssertionFailed(this.assertedCalls, this.callWriter, allCalls);
+            }
+

```

Comment:
Here's an example that would fail:

    A.CallTo(() => fake.Foo()).MustHaveHappened().InOrder();
    A.CallTo(() => fake.Bar()).MustHaveHappened().InOrder();
    A.CallTo(() => fake.Foo()).MustHaveHappened().InOrder();

The two calls to `Foo` would be consumed the first time, then we would miss the call to `Bar`.